modules/data.atmosphere/R/download.NOAA_GEFS.R

```diff
@@ -0,0 +1,260 @@
+    dir.create(outfolder, recursive=TRUE)
+  }
+  
+  # Create a data frame with information about the file.  This data frame's format is an internal PEcAn standard, and is stored in the BETY database to
+  # locate the data file.  The data file is stored on the local machine where the download occured.  Because NOAA GEFS is an 
+  # ensemble of 21 different forecast models, each model gets its own data frame.  All of the information is the same for 
+  # each file except for the file name.
+  results = data.frame(
+    file = "",                            #Path to the file (added in loop below).
+    host = PEcAn.remote::fqdn(),          #Name of the server where the file is stored
+    mimetype = "application/x-netcdf",    #Format the data is saved in
+    formatname = "CF Meteorology",        #Type of data
+    startdate = paste0(format(start_date, "%Y-%m-%dT%H:%M:00")),    #starting date and time, down to the second
+    enddate = paste0(format(end_date, "%Y-%m-%dT%H:%M:00")),        #ending date and time, down to the second
+    dbfile.name = "NOAA_GEFS",            #Source of data
+    stringsAsFactors = FALSE
+  )
+  
+  results_list = list()
+  
+  #Each ensemble gets its own file.
+  #These dimensions will be used for all 21 ncdf4 file members, so they're all declared once here.
+  #The data is really one-dimensional for each file (though we include lattitude and longitude dimensions
+  #to comply with the PEcAn standard).
+  time_dim = ncdf4::ncdim_def("Time", "6 Hours", 1:ncol(noaa_data[[1]]))
+  lat_dim = ncdf4::ncdim_def("Lattitude", "Degrees North", lat.in)
+  lon_dim = ncdf4::ncdim_def("Lattitude", "Degrees East", lon.in)
+  
+  dimensions_list = list(time_dim, lat_dim, lon_dim)
+  
+  nc_var_list = list()
+  for (i in 1:length(cf_var_names)) { #Each ensemble member will have data on each variable stored in their respective file.
+    nc_var_list[[i]] = ncdf4::ncvar_def(cf_var_names[i], cf_var_units[i], dimensions_list, missval=NaN)
+  }
+  
+  #For each ensemble
+  for (i in 1:21) { # i is the ensemble number
+    #Generating a unique identifier string that characterizes a particular data set.
+    identifier = paste("NOAA_GEFS", site_id, i, format(start_date, "%Y-%m-%dT%H:%M"), 
+          format(end_date, "%Y-%m-%dT%H:%M"), sep=".")
+    
+    ensemble_folder = file.path(outfolder, identifier)
+    
+    #Each file will go in its own folder.
+    if (!dir.exists(ensemble_folder)) {
+      dir.create(ensemble_folder, recursive=TRUE)
+    }
+    
+    flname = file.path(ensemble_folder, paste(identifier, "nc", sep = "."))
+    
+    #Each ensemble member gets its own unique data frame, which is stored in results_list
+    #Object references in R work differently than in other languages. When adding an item to a list, R creates a copy of it
+    #for you instead of just inserting the object reference, so this works.
+    results$file = flname
+    results_list[[i]] = results
+    
+    if (!file.exists(flname) | overwrite) {
```

Comment:
Inside of `if` statements, you almost always want `||` instead of `|`. The difference is that `||` (and its analogue `&&`) always return a single logical value and operate strictly from left to right, whereas `|` and `&` are vectorized across the entire expression. See [`?base::Logic`](http://stat.ethz.ch/R-manual/R-devel/library/base/html/Logic.html) for details.